### DISCUSSION TITLE
Where does this helpful error note about `#[axum::debug_handler]` come from?

### DISCUSSION THREAD

**Discussion by mikeyhew** [tokio-rs/axum#1615] [Q&A] [answered]
I noticed something pretty interesting in this error message I just got from the compiler about one of my handler functions:

```
error[E0277]: the trait bound `fn(RequestContext, Form<NewFeedForm>) -> impl futures::Future<Output = Result<impl IntoResponse, ServerError>> {handle_post_feeds}: Handler<_, _, _>` is not satisfied
   --> src/web/server.rs:37:31
    |
37  |         .route("/feeds", post(pages::new_feed::handle_post_feeds))
    |                          ---- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `Handler<_, _, _>` is not implemented for fn item `fn(RequestContext, Form<NewFeedForm>) -> impl futures::Future<Output = Result<impl IntoResponse, ServerError>> {handle_post_feeds}`
    |                          |
    |                          required by a bound introduced by this call
    |
    = note: Consider using `#[axum::debug_handler]` to improve the error message
    = help: the trait `Handler<T, S, B2>` is implemented for `axum::handler::Layered<L, H, T, S, B, B2>`
```

Specifically, this is the line that caught my attention:

```
= note: Consider using `#[axum::debug_handler]` to improve the error message
```

When I saw this line in the error message, the first thought that crossed my mind was, "there's no way they hard-coded a message about `axum::debug_handler` in the Rust compiler". Assuming that's the case — and for good measure, I ripgrepped through the [rust-lang/rust](https://github.com/rust-lang/rust) codebase to make sure — what secret magic is axum doing to tell the Rust compiler to include that note in the error message?

**Comment by davidpdrsn** [comment, marked as answer]
I'm happy you noticed that! We are using some nightly APIs to add those hints. See https://github.com/tokio-rs/axum/pull/1436

**Reply by mikeyhew** [reply]
That's really cool! I figured it was using some nightly features, but I didn't realize you could check for nightly in build.rs and enable a feature flag like that.

I also looked through the axum codebase before posting this and couldn't find anything related to this error message, but apparently it was added in 0.6 and I had the 0.5.x branch checked out for some reason 🤦‍♂️
